### PR TITLE
pybind: pickle the rados/rbd `Error` class

### PR DESCRIPTION
**Pull request by mgfritch**
the Error class does not have a `message` attribute:

```
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/cephadm/module.py", line 774, in process
    p.evaluate()
  File "/usr/share/ceph/mgr/cephadm/module.py", line 92, in evaluate
    self.finalize(None)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 559, in finalize
    self._first_promise._finalize(result)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 301, in _finalize
    self.fail(e)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 553, in fail
    super(Completion, self).fail(e)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 347, in fail
    self._exception = e
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 195, in _exception
    self._serialized_exception_ = pickle.dumps(e) if e is not None else None
  File "rados.pyx", line 374, in rados.Error.__reduce__
AttributeError: 'ObjectNotFound' object has no attribute 'message'
```

Fixes: https://tracker.ceph.com/issues/46852
Signed-off-by: Michael Fritch <mfritch@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
introduce by #15932

**Comment by sebastian-philipp** [comment]
also relates to #15574

File: src/pybind/rados/rados.pyx [hunk 1]
```diff
@@ -370,8 +370,6 @@ class Error(Exception):
             return msg
         return '[errno {0}] {1}'.format(self.errno, msg)
 
-    def __reduce__(self):
-        return (self.__class__, (self.message, self.errno))
```
Comment (sebastian-philipp) [inline src/pybind/rados/rados.pyx:374]: what about errno? How does this work now? Is this something new in py3?

**Comment by dillaman** [comment]
I am very confused by this PR. Unless the original need to support OpenStack and other legacy API users has been somehow satisfied by the switch the Python 3 -- I really don't feel comfortable.

**Comment by mgfritch** [comment]
> I am very confused by this PR. Unless the original need to support OpenStack and other legacy API users has been somehow satisfied by the switch the Python 3 -- I really don't feel comfortable.

The original need to support OpenStack et al. was satisfied by reversing the order of `message` and `errno` params in the ctor:

https://github.com/ceph/ceph/pull/15932/commits/9d47e5e4170bf9ddc3e642362027f7eb741380ca#diff-67abca628c07e7c03ac7310e67d32fe3L315-R316


> what about errno? How does this work now? Is this something new in py3?

Implementing a `__reduce__` function should not be needed because `errno` is passed via kwargs:
```
$ python2
Python 2.7.18 (default, Apr 23 2020, 09:27:04) [GCC] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> class Error(Exception):
...     def __init__(self, message, errno=None):
...         super(Exception, self).__init__(message)
...         self.errno = errno
...     def __str__(self):
...         msg = super(Exception, self).__str__()
...         if self.errno is None:
...             return msg
...         return '[errno {0}] {1}'.format(self.errno, msg)
... 

>>> e = Error('this is an error', errno=errno.EINVAL)
>>> e
Error('this is an error',)
>>> e.errno
22
>>> print(e)
[errno 22] this is an error
>>> e.__reduce__()
(<class '__main__.Error'>, ('this is an error',), {'errno': 22})

>>> import pickle
>>> s = pickle.dumps(e)
>>> e2 = pickle.loads(s)
>>> e2
Error('this is an error',)
>>> e2.errno
22
>>> print(e2)
[errno 22] this is an error
>>> e2.__reduce__()
(<class '__main__.Error'>, ('this is an error',), {'errno': 22})


```

**Comment by tchaikov** [comment]
@mgfritch hi Michael, could you revise the title of the commit message, 
as the title of the commit message reads like a bug report. ideally, the title of the commit message should use the imperative mood. see also https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes . please use `git commit --amend`  and `git push -f` for revising the commit message and its title.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-08-18_11:48:28-rados:cephadm-wip-swagner-testing-2020-08-18-1033-distro-basic-smithi/

Failure: https://tracker.ceph.com/issues/47009

**Comment by sebastian-philipp** [comment]
ping @dillaman ?

**Comment by dillaman** [comment]
> ping @dillaman ?

I don't feel comfortable enough to review this. I would need to see a test case added where these exceptions are serialized/deserialized (e.g. via `multiprocessing.Pipe.send/recv` which was the original source of the bug)

**Comment by sebastian-philipp** [comment]
pass: https://pulpito.ceph.com/swagner-2020-08-25_11:21:05-rados:cephadm-wip-swagner-testing-2020-08-25-1023-distro-basic-smithi/ 

(but that doesn't show anything really, as it passed before as well)

**Comment by sebastian-philipp** [comment]
this is a bit unsatisfying.

**Pull request closed without merging**